### ISSUE TITLE
[quiz] Show solutions of other participants

### ISSUE THREAD

**Issue by GaurangTandon** [firasdib/Regex101#1100]
## Feature

I solved my first Regex101 quiz question. In the stats it says there are people who took as many as 27 to 34 characters to solve the question, which surprised me as my solution was only 11 characters long. 

Thus, I wish to ask for a feature where it would let me see the solutions submitted by other people, as then I would learn of different possible interpretations of the same question, and/or develop a broad idea of what are the different ways to solve the same question.

This specially makes much more sense, when I have a longer solution, and would like to view shorter solutions of other participants. I mean, it is always possible for me to be able to brute force to the solution by manually applying all the characters, but looking at a simpler smaller elegant solution would be more of a learning opportunity.  
For example, in Task 7. Validate an IP, my solution is 113 lines long, while there are many popular shorter solutions of size 49, 75, and 93.

**Comment by Davidebyzero** [comment]
The problem with that is that if everybody could see everybody else's solutions, they would copy those and everybody would end up having the same shortest solution.

An example is [Regex Golf](https://alf.nu/RegexGolf), another site with regex puzzles (though very different than the regex101 quiz). It used to not show any scores of other people. This encouraged people to talk about it in online forums and discussion pages, and share their solutions freely. But then some years later, the site's creator revamped it, adding a visible ranked scores list with names and optional comments. Now, most of the levels are spammed with many copies of the best (lowest) score, and there's no way of knowing if those people came up with that solution independently, or just found one of the web page where the best published solutions are posted and copied them from there. And there's no longer any public discussion about various solutions to those levels.

**Comment by GaurangTandon** [comment]
Hello @Davidebyzero, thanks for your detailed reply. I very much appreciate what you wrote, and especially agree with you regarding the Regex Golf website. I also understand your sentiment that the shorter solutions should be more rewarding, and thereby getting to them easily would spoil the joy of actually getting to them.

Though, I would add that sometimes it's almost impossible to get to a shorter solution. Exactly as you said, for the life of me I cannot figure out how people did Task `#7` in less than half the characters as mine. I am surely ready to wait a bit longer, let it boil in my head for two to three weeks, google around for similar problems, and try different regex concepts in order to get to that answer. 

However, if even after one to two weeks of failed attempts a user doesn't get the solution, I think it should be fair enough to let them view the solution. I believe it provide a new learning opportunity for them. Maybe they never knew of this creative technique of using regex before. After viewing the solution they will be able to understand it better, and perhaps implement it on their own in future questions.

I'd add that I came to the quiz as a means to learn new stuff, and I feel I might not be able to learn as much given the current state. Surely I can brute force my way through all the quiz questions, but then it feels like one of those algorithm classes where every solution they teach is complexity `O(N^2)`. They are valid solutions, but they are terrible, and it is really the `O(NlogN)` and the `O(N)` which actually work, and which twist your brain to think in creative ways.

So, in short, I feel that we should figure out a proper implementation on how to show answers to quiz questions. Surely, giving everyone away every single solution is not what I'd recommend either. But, perhaps, after 30 failed attempts to a quiz question, over a period of one week, I should be able to "unlock" at least the next shorter solution (if not the shortest).

------

PS: I'd also add that it would be virtually impossible to keep the internet away from openly publishing solutions. As the quiz gets popular, sooner or later some blogger will take it up. Take Project Euler as an example. It's even stricter than Regex101 when it comes to checking answers (Regex101 at least gives hints as to what went wrong), but detailed solutions to it are available all over the internet.

**Comment by firasdib** [comment]
Personally I don't like revealing other peoples hard work, as there are a lot of clever people with clever solutions they might not want to share.

I also don't think the shortest solutions are the best, they are merely for golfing purposes. Many of them exploit the engine in certain ways which, quite frankly, make them awful for any real world use cases.

If you can pass the quiz, you have learned a lot. If you want to shorten your solutions, I recommend you join the discussions on IRC and ask for tips from participants there.

Just my two cents though.

**Comment by Davidebyzero** [comment]
**@GaurangTandon**, I'd love to talk with you about regex. I'm curious how you've done on the quiz since last posting here. Would you come over to #regex on freenode IRC and chat? (I have the same nickname there as here.)

**Comment by jpidelatorre** [comment]
Maybe the site could lock our responses when we uncover shorter answers. That way we would maintain our score and still learn new tricks. I'm eager to know how someone could validate an email in 45 characters.

@GaurangTandon The secret in #7 is to have an alternation to pick either a dot or the end of the match by the end of the expression. There's one thing, tho, you need a negative look ahead to ensure you don't have a dot right by the end of the match.

This is what I got after a few iterations: `^(?:(?:25[0-5]|2[0-4]\d|1?\d\d?)(?:\.(?!$)|$)){4}$` (52 chars). Someone managed to get the expression down to 40 characters but I really can't see where can I trim my expression down.

**Comment by Doqnach** [comment]
Please do not share solutions here. Please edit your post accordingly.

**Comment by firasdib** [comment]
What it feels this boils down to is that people want a forum for discussing the different solutions. I would suggest IRC for this. Perhaps we can create a room where users can discuss these things.

I don't think revealing the solutions on the website is good.

**Comment by jpidelatorre** [comment]
@firasdib I think we would need some place to have the discussions archived for future reference.

**Comment by firasdib** [comment]
I'll close this for now, but I would be happy to discuss the possibility of a regex101-forum :-)

**Comment by liudongmiao** [comment]
> This is what I got after a few iterations: (52 chars). Someone managed to get the expression down to 40 characters but I really can't see where can I trim my expression down.

@jpidelatorre I try to give a solution of 44 chars. However, I still don't know other solution, like 39.

**Comment by liudongmiao** [comment]
> This is what I got after a few iterations: (52 chars). Someone managed to get the expression down to 40 characters but I really can't see where can I trim my expression down.

@jpidelatorre I got a solution of 43 chars. However, I still don't know other solution, like 39.
And the best for AI, give best solution of 44 chars.
